pkg/endpoint/bpf.go

```diff
@@ -655,87 +662,134 @@ func (e *Endpoint) regenerateBPF(owner Owner, currentDir, nextDir string, regenC
 	os.RemoveAll(e.IPv6IngressMapPathLocked())
 	os.RemoveAll(e.IPv4IngressMapPathLocked())
+	return proxyCleanupFunc, nil
+}
+
+// regenerateBPF rewrites all headers and updates all BPF maps to reflect the
+// specified endpoint.
+// Must be called with endpoint.Mutex not held and endpoint.BuildMutex held.
+// Returns the policy revision number when the regeneration has called, a
+// boolean if the BPF compilation was executed and an error in case of an error.
+func (e *Endpoint) regenerateBPF(owner Owner, currentDir, nextDir string, regenContext *RegenerationContext) (revnum uint64, compiled bool, reterr error) {
+	stats := &regenContext.Stats
+	stats.waitingForLock.Start()
+
+	// Make sure that owner is not compiling base programs while we are
+	// regenerating an endpoint.
+	owner.GetCompilationLock().RLock()
+	defer owner.GetCompilationLock().RUnlock()
+
+	// Cache endpoint ID for use later.
+	e.UnconditionalRLock()
+	epID := e.StringID()
+	e.RUnlock()
+
+	bpfRegenContext := newDatapathRegenerationContext()
+	defer bpfRegenContext.completionCancel()
 
-	e.Unlock()
+	// Wrap logic which requires the endpoint to be locked within a function
+	// & call within regenerateBPF so that endpoint lock release can be deferred
+	// within the function. If this isn't done, and something else panics
+	// while the endpoint regeneration lock is held, and a deferred function
+	// within regenerateBPF tries to grab the endpoint lock, it will deadlock
+	// because the lock is already held, which obfuscates what happened in the
+	// initial panic.
+	// Returns an error, as well as a function which may need deferring
+	// based once this function returns. This function is intentionally not
+	// deferred in this function itself because it may acquire the endpoint
+	// lock.
+	revFunc, err := e.runPreCompilationSteps(owner, bpfRegenContext, currentDir, nextDir, regenContext)
+
+	// Note that revFunc may hold the endpoint lock.
+	if revFunc != nil {
+		defer revFunc(reterr)
+	}
+
+	if err != nil {
+		return 0, false, err
+	}
+
+	// No need to compile BPF in dry mode.
+	if option.Config.DryMode {
+		return e.nextPolicyRevision, false, nil
+	}
+
+	var compilationExecuted bool
 
 	// Wait for connection tracking cleaning to complete
 	stats.waitingForCTClean.Start()
-	<-ctCleaned
+	<-bpfRegenContext.ctCleaned
 	stats.waitingForCTClean.End(true)
 
-	e.getLogger().WithField("bpfHeaderfilesChanged", bpfHeaderfilesChanged).Debug("Preparing to compile BPF")
+	// The endpoint lock should not be held while compiling BPF, as this is a
+	// long-lived operation.
+	e.getLogger().WithField("bpfHeaderfilesChanged", bpfRegenContext.bpfHeaderfilesChanged).Debug("Preparing to compile BPF")
 
 	stats.prepareBuild.End(true)
-	if bpfHeaderfilesChanged || regenContext.ReloadDatapath {
-		closeChan := loadinfo.LogPeriodicSystemLoad(log.WithFields(logrus.Fields{logfields.EndpointID: epID}).Debugf, time.Second)
+	if bpfRegenContext.bpfHeaderfilesChanged || regenContext.ReloadDatapath {
```

Comment:
Maybe ReloadDatapath should be moved into `datapathRegenerationContext` - thoughts ?

Reply:
That's reasonable, I think. We can define the inner struct directly since it won't ever be referred to by type.

Reply:
Fixed in latest upload.